api/pull_request_test.go

```diff
@@ -2,42 +2,295 @@ package api
+		},
+		{
+			name:                 "SUCCESS is treated as Passing",
+			payload:              singleStatusContextWithState("SUCCESS"),
+			expectedChecksStatus: PullRequestChecksStatus{Passing: 1, Total: 1},
+		},
+		{
+			name:                 "ERROR is treated as Failing",
+			payload:              singleStatusContextWithState("ERROR"),
+			expectedChecksStatus: PullRequestChecksStatus{Failing: 1, Total: 1},
+		},
+		{
+			name:                 "FAILURE is treated as Failing",
+			payload:              singleStatusContextWithState("FAILURE"),
+			expectedChecksStatus: PullRequestChecksStatus{Failing: 1, Total: 1},
+		},
+		{
+			name:                 "Unrecognized States are treated as Pending",
+			payload:              singleStatusContextWithState("AnUnrecognizedStateJustForThisTest"),
+			expectedChecksStatus: PullRequestChecksStatus{Pending: 1, Total: 1},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			var pr PullRequest
+			require.NoError(t, json.Unmarshal([]byte(tt.payload), &pr))
+
+			require.Equal(t, tt.expectedChecksStatus, pr.ChecksStatus())
+		})
+	}
+}
+
+func TestChecksStatus_SummarisingCheckRunsAndStatusContexts(t *testing.T) {
+	t.Parallel()
+
+	// This might look a bit intimidating, but we're just inserting three nodes
+	// into the rollup, two completed check run nodes and one status context node.
+	payload := fmt.Sprintf(`
 	{ "statusCheckRollup": { "nodes": [{ "commit": {
 		"statusCheckRollup": {
 			"contexts": {
 				"nodes": [
-					{ "state": "SUCCESS" },
-					{ "state": "PENDING" },
-					{ "state": "FAILURE" },
-					{ "status": "IN_PROGRESS",
-					  "conclusion": null },
-					{ "status": "COMPLETED",
-					  "conclusion": "SUCCESS" },
-					{ "status": "COMPLETED",
-					  "conclusion": "FAILURE" },
-					{ "status": "COMPLETED",
-					  "conclusion": "ACTION_REQUIRED" },
-					{ "status": "COMPLETED",
-					  "conclusion": "STALE" }
+					%s,
+					%s,
+					%s
 				]
 			}
 		}
 	} }] } }
-	`
-	err := json.Unmarshal([]byte(payload), &pr)
-	assert.NoError(t, err)
-
-	checks := pr.ChecksStatus()
-	assert.Equal(t, 8, checks.Total)
-	assert.Equal(t, 3, checks.Pending)
-	assert.Equal(t, 3, checks.Failing)
-	assert.Equal(t, 2, checks.Passing)
+	`,
+		completedCheckRunNode("SUCCESS"),
+		statusContextNode("PENDING"),
+		completedCheckRunNode("FAILURE"),
+	)
+
+	var pr PullRequest
+	require.NoError(t, json.Unmarshal([]byte(payload), &pr))
+
+	expectedChecksStatus := PullRequestChecksStatus{
+		Pending: 1,
+		Failing: 1,
+		Passing: 1,
+		Total:   3,
+	}
+	require.Equal(t, expectedChecksStatus, pr.ChecksStatus())
+}
+
+// Note that it would be incorrect to provide a status of COMPLETED here
+// as the conclusion is always set to null. If you want a COMPLETED status,
+// use `singleCompletedCheckRunWithConclusion`.
+func singleCheckRunWithStatus(status string) string {
+	return fmt.Sprintf(`
+	{ "statusCheckRollup": { "nodes": [{ "commit": {
```

Comment:
For what it's worth, I _sort_ of feel like we should have a set of tests that validate deserialisation and the rest of the tests use structs directly but the anonymous structures we have in our types that back graphql flows aren't particularly nice to instantiate.

I'm not suggesting we make this change, I'm sure y'all have found advantages to starting from the string form, it just unsettles me a bit. Loosely held opinion!

Reply:
You are welcome to give all anonymous structs a name, but I'm not convinced that switching from a JSON representation of our test data to a pure Go representation would help test readability.